File: tests/fixtures/code_graph/ts/cache.ts
import { normalize } from "./util";

export interface Storable {
  key: string;
  hits: number;
}

export type CacheKey = string;

export abstract class BaseCache {
  protected entries: Map<CacheKey, Storable> = new Map();

  abstract label(): string;
}

export class Cache extends BaseCache implements Storable {
  key = "cache";
  hits = 0;

  label(): string {
    return "cache";
  }

  get(key: CacheKey): Storable | undefined {
    return this.entries.get(normalize(key));
  }

  put(entry: Storable): void {
    this.entries.set(normalize(entry.key), entry);
    this.hits += 1;
  }
}

export function buildCache(): Cache {
  return new Cache();
}

export const warmCache = (cache: Cache, keys: CacheKey[]): number => {
  let count = 0;
  for (const key of keys) {
    if (cache.get(key) === undefined) {
      cache.put({ key, hits: 0 });
      count += 1;
    }
  }
  return count;
};

File: tests/fixtures/code_graph/go/cache.go
package cache

import "strings"

const MaxEntries = 128

type Storable interface {
	Label() string
}

type Entry struct {
	Key  string
	Hits int
}

type Cache struct {
	entries map[string]Entry
}

func (c *Cache) Label() string {
	return "cache"
}

func (c *Cache) Get(key string) (Entry, bool) {
	entry, ok := c.entries[normalize(key)]
	return entry, ok
}

func (c *Cache) Put(entry Entry) {
	c.entries[normalize(entry.Key)] = entry
}

func BuildCache() *Cache {
	return &Cache{entries: make(map[string]Entry)}
}

func WarmCache(c *Cache, keys []string) int {
	count := 0
	for _, key := range keys {
		if _, ok := c.Get(key); !ok {
			c.Put(Entry{Key: key, Hits: 0})
			count++
		}
	}
	return count
}

func normalize(raw string) string {
	return strings.ToLower(strings.TrimSpace(raw))
}
